kernel/src/hil/rubble/rubble.rs

```diff
@@ -0,0 +1,157 @@
+//! the kernel. For example, these interfaces would be used to create a capsule
+//! that provides userspace with access to BLE via Rubble.
+
+use core::fmt;
+
+use crate::hil::time;
+use crate::ReturnCode;
+
+use super::types::{DeviceAddress, Duration, Instant, NextUpdate};
+
+/// The base trait representing for using the Rubble stack. The interface is
+/// based off of the types publicly exposed by Rubble which we use, and is
+/// otherwise fairly arbitrary.
+pub trait RubbleStack<'a, A: time::Alarm<'a> + ?Sized> {
+    type BleRadio: RubbleBleRadio<'a, A, Self>;
+    type LinkLayer: RubbleLinkLayer<'a, A, Self>;
+    type Responder: RubbleResponder<'a, A, Self>;
+    type Cmd: fmt::Debug + RubbleCmd;
+    /// The packet queue that this implementation uses.
+    type PacketQueue: RubblePacketQueue + 'static;
+
+    /// Retrieves this device's address.
+    fn get_device_address() -> DeviceAddress;
+
+    /// Retrieve a statically-allocated packet queue for communication from the
+    /// hardware to the LinkLayer.
+    fn rx_packet_queue() -> &'static Self::PacketQueue;
+
+    /// Retrieve a statically-allocated packet queue for communication from the
+    /// LinkLayer to the hardware.
+    fn tx_packet_queue() -> &'static Self::PacketQueue;
+
+    /// Handle a receive event from the BLE stack. This is here as it needs
+    /// access to the BleRadio and LinkLayer structs simultaneously.
+    fn transmit_event(
+        radio: &mut Self::BleRadio,
+        ll: &mut Self::LinkLayer,
+        rx_end: Instant,
+        buf: &'static mut [u8],
+        result: ReturnCode,
+    );
+
+    /// Handle a receive event from the BLE stack. This is here as it needs
+    /// access to the BleRadio and LinkLayer structs simultaneously.
+    fn receive_event(
+        radio: &mut Self::BleRadio,
+        ll: &mut Self::LinkLayer,
+        rx_end: Instant,
+        buf: &'static mut [u8],
+        len: u8,
+        result: ReturnCode,
+    ) -> Self::Cmd;
+}
+
+pub trait RubbleBleRadio<'a, A, R>
+where
+    A: time::Alarm<'a> + ?Sized,
+    R: RubbleStack<'a, A> + ?Sized,
+{
+    fn accept_cmd(&mut self, cmd: <R::Cmd as RubbleCmd>::RadioCmd);
+}
+
+pub trait RubbleLinkLayer<'a, A, R>
+where
+    A: time::Alarm<'a> + ?Sized,
+    R: RubbleStack<'a, A> + ?Sized,
+{
+    type Error: fmt::Debug + fmt::Display;
+    fn new(device_address: DeviceAddress, alarm: &'a A) -> Self;
+
+    /// Starts advertising this device, optionally sending data along with the
+    /// advertising PDU.
+    ///
+    /// # Errors
+    ///
+    /// This will error if the advertising data is incorrectly formatted, or if
+    /// it's too long.
+    fn start_advertise(
+        &mut self,
+        interval: Duration,
+        data: &[u8],
+        transmitter: &mut R::BleRadio,
+        tx: <R::PacketQueue as RubblePacketQueue>::Consumer,
+        rx: <R::PacketQueue as RubblePacketQueue>::Producer,
+    ) -> Result<NextUpdate, Self::Error>;
```

Comment:
Here is the interface definition for the interface tock capsules use for the Rubble stack to start advertising. This returns `NextUpdate` which the caller has to then keep track of make sure there is a timer set to fire at that time. Having operations return the time to do the next operation is how Rubble organizes events, but that seems like it should be contained in the tock-rubble adapter, and hidden from the tock capsules.

Reply:
Ah I see.

I'm guessing it's because the tock_rubble support has a few different layers (as different structs/traits) so there is no single alarm instance. In this case I think it is cleaner to just pass the `NextUpdate` to the capsule where it can all be managed in one place.

As well as that doesn't it match other capsules in Tock where the capsule is in charge of setting alarms? It also means a different capsule could in theory do different operations with the `NextUpdate`.

Reply:
Why is that managed in one place? That is very much a Rubble-only concept, and the Rubble management is then split between the tock-rubble crate and any user of BLE (which could be several capsules). It's even worse if multiple kernel capsules want to use BLE (say for app updates and to provide BLE to userspace).

Reply:
Managed in one place in the sense that there is a single alarm function.

An alarm for a BLE implementation is common and not just a Rubble concept. I would expect any BLE capsule to use an alarm.

I see what you are saying though, if we could move it out of the capsule we can have less duplication between multiple BLE related capsules.

Maybe that can be changed later? These is still lots missing from this implementation so I we'll still need more work here.